server/src/main/java/io/crate/execution/engine/join/HashJoinBatchIterator.java

```diff
@@ -165,6 +197,34 @@ public boolean moveNext() {
         return true;
     }
 
+    private void extractNonMatchingKeys() {
+        if (nonMatchingKeys == null) {
+            nonMatchingKeys = new ArrayList<>();
+            for (var values : buffer.entrySet()) {
+                if (values.getValue().marked == false) {
+                    nonMatchingKeys.add(values.getKey());
+                }
+            }
+        }
+    }
+
+    private boolean emitNullValuesPairs() {
+        if (nonMatchValuesIterator == null) {
+            var key = nonMatchingKeys.getFirst();
+            nonMatchValuesIterator = buffer.get(key).items.iterator();
+        }
+
+        combiner.setLeft(unsafeArrayRow.cells(nonMatchValuesIterator.next()));
+        combiner.nullRight();
+
+        if (nonMatchValuesIterator.hasNext() == false) {
+            nonMatchingKeys.removeFirst();
```

Comment:
This uses `System.arrayCopy()` for `ArrayList`, so I'd use a `LinkedList` instead where removal of elements is much cheaper. You could verify  if this helps with a jmh test.

Reply:
Another variant to try: an `IntDeque` and change the `buffer.entrySet()` loop above to use `entries` with the cursor, to avoid `int` -> `Integer` boxing.

Reply:
👍